src/components/PostDetails.tsx

```diff
@@ -1,106 +1,184 @@
-import React from 'react';
+import React, { useEffect, useState } from 'react';
 import { Loader } from './Loader';
 import { NewCommentForm } from './NewCommentForm';
+import { Post } from '../types/Post';
+import { delateComment, getComments, sendComment } from '../api/comment';
+import { Comment } from '../types/Comment';
+
+type Props = {
+  currentPost: Post;
+};
+
+export const PostDetails: React.FC<Props> = ({ currentPost }) => {
+  const [isCommentsLoader, setIsCommentsLoader] = useState(false);
+  const [hasError, setHasError] = useState(false);
+  const [hasNewCommentForm, setHasNewCommentForm] = useState(false);
+  const [comments, setComments] = useState<Comment[]>([]);
+
+  useEffect(() => {
+    if (!currentPost) {
+      return;
+    }
+
+    setIsCommentsLoader(true);
+    setHasNewCommentForm(false);
+
+    getComments(currentPost.id)
+      .then(commentsFromServer => {
+        setComments(commentsFromServer);
+      })
+      .catch(() => setHasError(true))
+      .finally(() => setIsCommentsLoader(false));
+  }, [currentPost]);
+
+  const removeCommentFromServer = (
+    commentId: number,
+    index: number,
+    forDelete: Comment,
+  ) => {
+    delateComment(commentId).catch(() =>
+      setComments(prevComments => {
+        const workComments = [...prevComments];
+
+        workComments.splice(index, 1, forDelete);
+
+        return workComments;
+      }),
+    );
+  };
+
+  const handleDelete = (commentId: number) => {
+    const index = comments.findIndex(c => c.id === commentId);
+    const forDelete = { ...comments[index] };
+
+    setComments(prevComments => {
+      const workComments = [...prevComments];
+
+      workComments.splice(index, 1);
+
+      return workComments;
+    });
+
+    if (commentId === 0) {
+      return;
+    }
+
+    removeCommentFromServer(commentId, index, forDelete);
+  };
+
+  const addCommentOnServer = (newComment: Comment) => {
+    setComments(prevComments => [...prevComments, newComment]);
+
+    return sendComment(newComment)
+      .then(newFromServerComment => {
+        setComments(prevComments => {
+          const index = prevComments.findIndex(c => c.id === newComment.id);
+
+          if (index === -1) {
+            delateComment(newFromServerComment.id);
+
+            return prevComments;
+          }
+
+          const workComments = [...prevComments];
+
+          workComments.splice(index, 1, newFromServerComment);
+
+          return workComments;
+        });
+      })
+      .catch(() => {
+        setComments(prevComments => {
+          const index = prevComments.findIndex(c => c.id === 0);
+
+          const workComments = [...prevComments];
+
+          workComments.splice(index, 1);
+
+          return workComments;
+        });
```

Comment:
do not touch the comments on error, just tell the user that something is went wrong